x-pack/metricbeat/module/gcp/billing/billing.go

```diff
@@ -266,34 +302,63 @@ func (m *MetricSet) queryBigQuery(ctx context.Context, client *bigquery.Client,
 			return events, err
 		}
 
-		if len(row) == 6 {
-			events = append(events, createEvents(row, m.config.ProjectID))
-		}
+		events = append(events, createEvents(row, tableMeta.tableFullID, m.config.ProjectID))
 	}
 	return events, nil
 }
 
-func createEvents(rowItems []bigquery.Value, projectID string) mb.Event {
+// createTagsMap converts a comma-separated string of key-value pairs into a flattened map.
+// It is used to convert the tags string generated by the SQL query to a map.
+// The SQL query generates tags string by concatenating key-value pairs from an array of structs (tags).
+func createTagsMap(tagsItem bigquery.Value) mapstr.M {
+	tagsMap := mapstr.M{}
+
+	if tags, ok := tagsItem.(string); ok {
+		pairs := strings.Split(tags, ",")
+		for _, pair := range pairs {
+			kv := strings.Split(pair, ":")
+			if len(kv) == 2 {
+				_, _ = tagsMap.Put(kv[0], kv[1])
+			}
+		}
+	}
+
+	return tagsMap.Flatten()
+}
+
+func createEvents(row row, tableName, projectID string) mb.Event {
 	event := mb.Event{}
+
+	tagsMap := createTagsMap(row.Tags)
+
 	event.MetricSetFields = mapstr.M{
-		"invoice_month":      rowItems[0],
-		"project_id":         rowItems[1],
-		"project_name":       rowItems[2],
-		"billing_account_id": rowItems[3],
-		"cost_type":          rowItems[4],
-		"total":              rowItems[5],
+		"invoice_month":       row.InvoiceMonth,
+		"project_id":          row.ProjectId,
+		"project_name":        row.ProjectName,
+		"billing_account_id":  row.BillingAccountId,
+		"cost_type":           row.CostType,
+		"total":               row.TotalExact,
+		"sku_id":              row.SkuId,
+		"sku_description":     row.SkuDescription,
+		"service_id":          row.ServiceId,
+		"service_description": row.ServiceDescription,
+		"tags":                tagsMap,
+	}
+
+	if isDetailedTable(tableName) {
+		_, _ = event.MetricSetFields.Put("effective_price", row.EffectivePrice)
```

Comment:
Shouldn't we also add all the new fields to the mapping?

Reply:
Oh, I guess this is already covered by the existing mapping:

```yaml
        - name: "metrics.*.*.*.*"
          type: object
          object_type: double
          object_type_mapping_type: "*"
          description: >
            Metrics that returned from Google Cloud API query.
```

right?

Reply:
Yeah, will add the fields to the mapping!